File: java/algorithms/union_find.java
class AUnionFind {
    static int find(int[] p,int x){ return p[x]==x?x:(p[x]=find(p,p[x])); }
    static void union(int[] p,int a,int b){ a=find(p,a); b=find(p,b); if(a!=b)p[b]=a; }
    public static void main(String[] args){
        int[] p=new int[6]; for(int i=0;i<p.length;i++) p[i]=i;
        union(p,1,2); union(p,2,3); union(p,4,5);
        System.out.println(find(p,1)==find(p,3));
        System.out.println(find(p,1)==find(p,5));
    }
}

File: c/algorithms/union_find.c
#include <stdio.h>

int parent[10];

int find(int x){
    if(parent[x]==x) return x;
    return parent[x]=find(parent[x]);
}

void unite(int a,int b){
    a=find(a); b=find(b);
    if(a!=b) parent[b]=a;
}

int main(void){
    for(int i=0;i<10;i++) parent[i]=i;
    unite(1,2); unite(2,3);
    printf("1과 3 같은 집합? %s\n", find(1)==find(3)?"YES":"NO");
    return 0;
}
